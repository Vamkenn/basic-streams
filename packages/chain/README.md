# @basic-streams/chain

TODO: fill up README
