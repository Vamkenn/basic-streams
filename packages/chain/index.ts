import {Stream} from "@basic-streams/stream"

export default function chain<T, U>(
  fn: (x: T) => Stream<U>,
  stream: Stream<T>,
): Stream<U> {
  return cb => {
    let spawnedDisposers: Array<() => void> = []
    const mainDisposer = stream(payload => {
      spawnedDisposers.push(fn(payload)(cb))
    })
    return () => {
      spawnedDisposers.forEach(fn => fn())
      mainDisposer()
    }
  }
}
