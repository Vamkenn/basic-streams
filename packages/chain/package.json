{
  "name": "@basic-streams/chain",
  "version": "0.0.7",
  "description": "chain operator for basic-streams",
  "keywords": [
    "basic-streams",
    "chain"
  ],
  "author": "Roman Pominov <rpominov@gmail.com>",
  "main": "./index.js",
  "types": "./index.ts",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/rpominov/basic-streams.git"
  },
  "bugs": {
    "url": "https://github.com/rpominov/basic-streams/issues"
  },
  "homepage": "https://github.com/rpominov/basic-streams/packages/chain",
  "dependencies": {
    "@basic-streams/stream": "^0.0.5"
  }
}
