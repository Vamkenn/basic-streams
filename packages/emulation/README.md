# [@basic-streams](https://github.com/rpominov/basic-streams)/emulation

Although the version is above `1.0` this package is not stable. Please use with
caution API may change.
