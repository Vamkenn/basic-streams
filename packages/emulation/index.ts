import {Stream} from "@basic-streams/stream"

export class Value<T> {
  constructor(public readonly value: T) {}
}

export class TimeSpan {
  constructor(public readonly ms: number) {}
}

export type Timeline<T> = Array<TimeSpan | Value<T>>

export class Event<T> {
  constructor(
    public readonly time: number,
    public readonly value: T,
    public readonly cb?: ((x: T) => void),
  ) {}

  callCb() {
    if (this.cb) {
      this.cb(this.value)
    }
  }
}

export class EventsList<T> {
  constructor(public readonly items: Array<Event<T>>) {}

  static fromTimeline<T>(items: Timeline<T>, currentTime = 0): EventsList<T> {
    const events: Array<Event<T>> = []
    for (const item of items) {
      if (item instanceof TimeSpan) {
        currentTime = currentTime + item.ms
      } else {
        events.push(new Event(currentTime, item.value))
      }
    }
    return new EventsList(events)
  }

  withCb(cb: (value: T) => void): EventsList<T> {
    return new EventsList(
      this.items.map(item => new Event(item.time, item.value, cb)),
    )
  }

  merge(another: EventsList<T>): EventsList<T> {
    const events: Array<Event<T>> = []
    let indexA = 0
    let indexB = 0
    const itemsA = this.items
    const itemsB = another.items
    while (itemsA.length > indexA || itemsB.length > indexB) {
      if (
        itemsA.length !== indexA &&
        (itemsB.length === indexB || itemsA[indexA].time <= itemsB[indexB].time)
      ) {
        events.push(itemsA[indexA])
        indexA++
      } else {
        events.push(itemsB[indexB])
        indexB++
      }
    }
    return new EventsList(events)
  }

  takeOne(): {event: Event<T> | void; rest: EventsList<T>} {
    const [event, ...rest] = this.items
    return {event, rest: new EventsList(rest)}
  }

  toJSON() {
    return this.items.map(item => ({time: item.time, value: item.value}))
  }

  static jestSerializer = {
    test(x: any) {
      return x instanceof EventsList
    },

    // for outdated TS typings
    print(x: any): string {
      return ""
    },

    serialize<Refs, Config extends {indent: string}, T>(
      val: EventsList<T>,
      config: Config,
      indentation: string,
      depth: number,
      refs: Refs,
      printer: (
        x: T,
        config: Config,
        indentation: string,
        depth: number,
        refs: Refs,
      ) => string,
    ) {
      const separator = "\n"

      function printItem(item: Event<T>) {
        const indentation1 = indentation + config.indent
        const value = printer(item.value, config, indentation1, depth + 1, refs)
        return indentation1 + item.time + ": " + value
      }

      return `EventsList(${
        val.items.length === 0
          ? ""
          : separator + val.items.map(printItem).join(separator) + separator
      })`
    },
  }
}

export function t(ms: number): TimeSpan {
  return new TimeSpan(ms)
}

export function v<T>(x: T): Value<T> {
  return new Value(x)
}

export function emulate<T>(
  generator: (
    createStream: <U>(...timeline: Timeline<U>) => Stream<U>,
  ) => Stream<T>,
  maxTime = Infinity,
): EventsList<T> {
  const state = {
    time: 0,
    toProduce: new EventsList([]) as EventsList<any>,
    result: [] as Array<Event<T>>,
  }
  const resultStream = generator((...timeline) => {
    return cb => {
      // wrapping cb to make it a unique value
      // that we will compare to in unsuscribe
      const _cb = (x: any) => cb(x)
      state.toProduce = state.toProduce.merge(
        EventsList.fromTimeline(timeline, state.time).withCb(_cb),
      )
      return () => {
        state.toProduce = new EventsList(
          state.toProduce.items.filter(item => item.cb !== _cb),
        )
      }
    }
  })
  resultStream(value => {
    state.result.push(new Event(state.time, value))
  })
  while (true) {
    const {event, rest} = state.toProduce.takeOne()
    if (state.time >= maxTime || !event) {
      return new EventsList(state.result)
    }
    state.time = event.time
    state.toProduce = rest
    event.callCb()
  }
}

export function laterMock(
  createStream: <U>(...timeline: Timeline<U>) => Stream<U>,
) {
  return function later<T>(time: number, value?: T): Stream<T> {
    return createStream(t(time), v(value as any))
  }
}
