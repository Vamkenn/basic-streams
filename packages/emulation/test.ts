import {EventsList, emulate, t, v} from "./index"

expect.addSnapshotSerializer(EventsList.jestSerializer)

describe("emulate", () => {
  test("map", () => {
    expect(
      emulate(create => {
        return cb => create(t(10), v(1))(x => cb(x + 1))
      }),
    ).toMatchSnapshot()
  })

  test("filter", () => {
    expect(
      emulate(create => {
        return cb =>
          create(v(1), t(10), v(0), t(10), v(1))(x => {
            if (x > 0) cb(x)
          })
      }),
    ).toMatchSnapshot()
  })

  test("order of simultaneous values", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1))
        const b = create(t(10), v(2))
        return cb => {
          const disposeA = a(cb)
          const disposeB = b(cb)
          return () => {
            disposeA()
            disposeB()
          }
        }
      }),
    ).toMatchSnapshot()
  })

  test("unsubscribe", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1), t(10), v(0))
        const b = create(t(15), v(null))
        return cb => b(a(cb))
      }),
    ).toMatchSnapshot()
  })

  test("unsubscribe doesn't remove other values with same cb", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1), t(10), v(0))
        const b = create(t(10), v(1), t(10), v(0))
        const unsubscriber = create(t(15), v(null))
        return cb => {
          a(cb)
          unsubscriber(b(cb))
          return () => {}
        }
      }),
    ).toMatchSnapshot()
  })
})
