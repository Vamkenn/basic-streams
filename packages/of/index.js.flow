// @flow

import type {Stream} from "@basic-streams/stream"

declare export default function of<T>(T): Stream<T>
