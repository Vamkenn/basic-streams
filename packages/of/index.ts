import {Stream} from "@basic-streams/stream"

function noop() {}

export default function of<T>(x: T): Stream<T> {
  return cb => {
    cb(x)
    return noop
  }
}
