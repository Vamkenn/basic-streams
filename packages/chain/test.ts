import {EventsList, emulate, t, v} from "@basic-streams/emulation"
import chain from "./index"
import of from "../of/index"

expect.addSnapshotSerializer(EventsList.jestSerializer)

function noop() {}

test("preserves disposer of main sream", () => {
  const disposer = jest.fn()
  chain(x => of(x), cb => disposer)(noop)()
  expect(disposer.mock.calls).toMatchSnapshot()
})

test("preserves disposer of spawned streams", () => {
  const disposer = jest.fn()
  chain(() => cb => disposer, of(null))(noop)()
  expect(disposer.mock.calls).toMatchSnapshot()
})

test("result stream contains values from spawned at correct time", () => {
  const result = emulate(create => {
    return chain(
      x => create(t(1), v(x * 2), t(1), v(x * 3)),
      create(t(10), v(1), t(10), v(2)),
    )
  })
  expect(result).toMatchSnapshot()
})
