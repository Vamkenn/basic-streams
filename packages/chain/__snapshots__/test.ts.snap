// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preserves disposer of main sream 1`] = `
Array [
  Array [],
]
`;

exports[`preserves disposer of spawned streams 1`] = `
Array [
  Array [],
]
`;

exports[`result stream contains values from spawned at correct time 1`] = `
EventsList(
  11: 2
  12: 3
  21: 4
  22: 6
)
`;
